# ELASPIC

[![conda](https://img.shields.io/conda/dn/kimlab/elaspic.svg)](https://anaconda.org/kimlab/elaspic/)
[![pipeline status](https://gitlab.com/elaspic/elaspic/badges/master/pipeline.svg)](https://gitlab.com/elaspic/elaspic/commits/master/)
[![coverage report](https://gitlab.com/elaspic/elaspic/badges/master/coverage.svg?job=test-unittest)](https://gitlab.com/elaspic/elaspic/commits/master/)

## Introduction

Welcome to the ELASPIC code repository!

Complete documentation is available on [ReadTheDocs](https://elaspic.readthedocs.io).

For a small number of mutations, you can try running ELASPIC using our [webserver](http://elaspic.kimlab.org).

## Development

- Use the issue tracker to discuss changes.

## References

Strokach, A., Kim, P.M. (2016) [_Predicting the Effect of Mutations on a Genome-Wide Scale._](https://github.com/ostrokach/msc_thesis/blob/master/msc_thesis.pdf) MSc Thesis, University of Toronto.

Witvliet, D., Strokach, A., Giraldo-Forero, A.F., Teyra, J., Colak, R., and Kim, P.M. (2016)
_ELASPIC web-server: proteome-wide structure based prediction of mutation effects on protein stability and binding affinity._ Bioinformatics 32(10): 1589-1591. doi: [10.1093/bioinformatics/btw031](https://doi.org/10.1093/bioinformatics/btw031).

Berliner N, Teyra J, Ã‡olak R, Garcia Lopez S, Kim PM (2014) _Combining Structural Modeling with Ensemble Machine Learning to Accurately Predict Protein Fold Stability and Binding Affinity Effects upon Mutation._ PLoS ONE 9(9): e107353. doi: [10.1371/journal.pone.0107353](https://doi.org/10.1371/journal.pone.0107353).

## License

MIT
